zkevm-circuits/src/evm_circuit/util.rs

```diff
@@ -542,30 +528,45 @@ pub(crate) mod from_bytes {
 /// Returns the random linear combination of the inputs.
 /// Encoding is done as follows: v_0 * R^0 + v_1 * R^1 + ...
 pub(crate) mod rlc {
+    use std::ops::{Add, Mul};
+
     use crate::util::Expr;
     use halo2_proofs::{arithmetic::FieldExt, plonk::Expression};
 
-    pub(crate) fn expr<F: FieldExt, E: Expr<F>>(
-        expressions: &[E],
-        power_of_randomness: &[E],
-    ) -> Expression<F> {
-        debug_assert!(expressions.len() <= power_of_randomness.len() + 1);
-
-        let mut rlc = expressions[0].expr();
-        for (expression, randomness) in expressions[1..].iter().zip(power_of_randomness.iter()) {
-            rlc = rlc + expression.expr() * randomness.expr();
+    pub(crate) fn expr<F: FieldExt, E: Expr<F>>(expressions: &[E], randomness: E) -> Expression<F> {
+        if !expressions.is_empty() {
+            gen(expressions.iter().map(|e| e.expr()), randomness.expr())
+        } else {
+            0.expr()
         }
-        rlc
     }
 
     pub(crate) fn value<'a, F: FieldExt, I>(values: I, randomness: F) -> F
     where
         I: IntoIterator<Item = &'a u8>,
         <I as IntoIterator>::IntoIter: DoubleEndedIterator,
     {
-        values.into_iter().rev().fold(F::zero(), |acc, value| {
-            acc * randomness + F::from(*value as u64)
-        })
+        let values = values
+            .into_iter()
+            .map(|v| F::from(*v as u64))
+            .collect::<Vec<F>>();
+        if !values.is_empty() {
+            gen(values, randomness)
+        } else {
+            F::zero()
+        }
+    }
+
+    pub(crate) fn gen<V, I>(values: I, randomness: V) -> V
```

Comment:
If I could use a trait that includes ::zero then I could return it when the values is an empty iterator. Now, this is only done for expr and value, I am thinking in creating an issue in FF.

Reply:
You mean defining the trait `Field` in `halo2_proofs::arithmetic::Field` as `... + num_traits::identities::Zero` right?

Reply:
I have been looking into this, I think https://crates.io/crates/num-traits is not standard, and we would need to convinced several open source projects to use that trait.

I think it is a pitty that core::ops does not include a Zero and One trait, perhaps we should request it.